### DISCUSSION TITLE
Typer options unexpectedly being propagated to pytest

### DISCUSSION THREAD

**Discussion by virgesmith** [fastapi/typer#817] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the Typer documentation, with the integrated search.
- [X] I already searched in Google "How to X in Typer" and didn't find any information.
- [X] I already read and followed all the tutorials in the docs and didn't find an answer.
- [X] I already checked if it is not related to Typer but to [Click](https://github.com/pallets/click).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import typer
import pytest


def main(some_opt: str = typer.Option(default="default")):
    pytest.main()


if __name__ == "__main__":
    typer.run(main)
```

### Description

Above code saved to `issue.py` If - and only if - the option is specified, it is somehow propagated to pytest. This works as expected (runs pytest):
```
python issue.py
```
but this fails:
```
python issue.py --some-opt X

ERROR: usage: issue.py [options] [file_or_dir] [file_or_dir] [...]
issue.py: error: unrecognized arguments: --some-opt
  inifile: /home/.../pyproject.toml
  rootdir: /home/...
```
which is a pytest error message with `pytest` being replaced with `issue.py`, which you can confirm by running
```
pytest --some-opt X
ERROR: usage: pytest [options] [file_or_dir] [file_or_dir] [...]
pytest: error: unrecognized arguments: --some-opt
  inifile: /home/.../pyproject.toml
  rootdir: /home/...
```

### Operating System

Linux

### Typer Version

0.12.3

### Python Version

3.12.2

**Comment by xandaaah** [comment]
Here, I fixed it for you.
```
import typer
import sys
import copy
argv = copy.deepcopy(sys.argv) # If you say a = 1, b = a, and b = 2, a will equal 2. Copies the variable to a different section of memory, keeping the original variable untouched.
sys.argv = [argv[0]] # Only add the python file name to argv.
import pytest
sys.argv = argv # Restore the argv so Typer can recognize it



def main(some_opt: str = typer.Option(default="default")):
    pytest.main()


if __name__ == "__main__":
    typer.run(main)
```
This works by tricking PyTest into thinking it got no arguments, so Typer can react to them.

**Comment by virgesmith** [comment]
I found a much better approach. Since `pytest.main` optionally accepts args that override `sys.argv`, I don't need to actually modify it. I just need to modify the code above to tell typer not to complain about any args it's not expecting, and also store them for me (as documented [here](https://typer.tiangolo.com/tutorial/commands/context/#configuring-the-context))


```py
import typer
import pytest


@app.command(context_settings={"allow_extra_args": True, "ignore_unknown_options": True})
def main(ctx: typer.Context, some_opt: str = typer.Option(default="default")):
    # do stuff with some_opt
    ...
    pytest.main(ctx.args)


if __name__ == "__main__":
    typer.run(main)
```

So now if I call

```sh
python issue.py -s tests/test_things.py --some-opt X
```

it will do the thing with the option, and then actually call

```sh
pytest -s tests/test_things.py
```

NB For typer arguments (as opposed to options), the ordering becomes important. But this isn't an issue for me.
